org/mozilla/jss/ssl/javax/JSSSocketChannel.java

```diff
@@ -0,0 +1,495 @@
+    /**
+     * Compute the total size of a list of buffers from the specified offest
+     * and length.
+     */
+    private static long computeSize(ByteBuffer[] buffers, int offset, int length) throws IOException {
+        long result = 0;
+
+        if (buffers == null || buffers.length == 0) {
+            return result;
+        }
+
+        for (int rel_index = 0; rel_index < length; rel_index++) {
+            int index = offset + rel_index;
+            if (index >= buffers.length) {
+                String msg = "Offset (" + offset + " or length (" + length;
+                msg += ") exceeds contract based on number of buffers ";
+                msg += "given (" + buffers.length + ")";
+                throw new IOException(msg);
+            }
+
+            if (buffers[index] != null) {
+                result += buffers[index].remaining();
+            }
+        }
+
+        return result;
+    }
+
+    public int read(ByteBuffer dst) throws IOException {
+        return (int) read(new ByteBuffer[] { dst });
+    }
+
+    public synchronized long read(ByteBuffer[] dsts, int offset, int length) throws IOException {
+        if (inboundClosed) {
+            return -1;
+        }
+
+        readBuffer.clear();
+
+        int buffer_size = boundRead(readBuffer.capacity());
+        ByteBuffer src = ByteBuffer.wrap(readBuffer.array(), 0, buffer_size);
+
+        long remoteRead = 0;
+        if (consumed != null) {
+            remoteRead = consumedChannel.read(src);
+        } else {
+            remoteRead = readChannel.read(src);
+        }
+
+        if (remoteRead == 0) {
+            return 0;
+        }
+
+        src.flip();
+
+        long unwrapped = 0;
+        long decrypted = 0;
+
+        try {
+            do {
+                SSLEngineResult result = engine.unwrap(src, dsts, offset, length);
+                if (result.getStatus() != SSLEngineResult.Status.OK && result.getStatus() != SSLEngineResult.Status.CLOSED) {
+                    throw new IOException("Unexpected status from unwrap: " + result);
+                }
+
+                unwrapped += result.bytesConsumed();
+                decrypted += result.bytesProduced();
+
+                if (unwrapped < remoteRead && result.bytesConsumed() == 0 && result.bytesProduced() == 0) {
+                    String msg = "Error attempting to read from remote peer: ";
+                    msg += "calls to unwrap stalled, consuming and producing ";
+                    msg += "no data: unwrapped " + unwrapped + " bytes of ";
+                    msg += remoteRead + " bytes read from remote peer; got a ";
+                    msg += "decrypted size of " + decrypted + " bytes.";
+                    throw new IOException(msg);
+                }
+            } while (unwrapped < remoteRead);
+        } catch (SSLException ssle) {
+            String msg = "Error attempting to read from remote peer: ";
+            msg += "got unexpected exception during unwrap call: ";
```

Comment:
The exception message probably can be shortened like "Unable to unwrap SSL data:".